File: python/chainpack/test_rpcmessage.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import pytest
from rpcmessage import *

def testRpcRequest():
	print("============= chainpack rpcmessage test ============\n")
	print("------------- RpcRequest")
	rq = RpcRequest()
	rq.setId(123)
	rq.setMethod("foo")
	rq.setParams({"a": 45,
	            "b": "bar",
				"c": RpcValue([1,2,3])})
	rq.setMetaValue(meta.RpcMessage.Tag.DeviceId, "aus/mel/pres/A");
	cp1 = rq._value
	out = ChainPackProtocol()
	len = rq.write(out);
	cp2: RpcValue = out.read()
	print (cp1, " " ,cp2," len: ", len," dump: " ,out);
	assert cp1.type == cp2.type
	rq2 = RpcRequest(cp2)
	assert rq2.isRequest()
	assert rq2.id() == rq.id()
	assert rq2.method() == rq.method()
	assert rq2.params() == rq.params()

def testRpcResponse():
	print("------------- RpcResponse")
	rs = RpcResponse()
	rs.setId(123)
	rs.setResult(42);
	out = ChainPackProtocol()
	cp1: RpcValue = rs._value
	len: int = rs.write(out)
	cp2: RpcValue = out.read()
	print(cp1, " ", cp2, " len: ", len, " dump: ", str(out));
	assert cp1.type == cp2.type
	rs2: RpcResponse = RpcResponse(cp2);
	assert rs2.isResponse()
	assert rs2.id() == rs.id()
	assert rs2.result() == rs.result()

def testRpcResponse2():
	rs = RpcResponse()
	rs.setId(123)
	rs.setError(RpcResponse.Error.createError(RpcResponse.ErrorType.InvalidParams, "Paramter length should be greater than zero!"))
	out = ChainPackProtocol()
	cp1: RpcValue = rs._value;
	l: int = rs.write(out);
	cp2: Value = out.read()
	print(cp1, " ", cp2, " len: ", l, " dump: ", out);
	assert(cp1.type == cp2.type);
	rs2 = RpcResponse(cp2);
	assert(rs2.isResponse());
	assert rs2.id() == rs.id()
	assert rs2.error() == rs.error()

def testRpcNotify():
	print("------------- RpcNotify")
	rq = RpcRequest()
	rq.setMethod("foo")
	rq.setParams({"a": 45,
				  "b": "bar",
				  "c": RpcValue([1,2,3])});
	out = ChainPackProtocol()
	cp1: RpcValue = rq._value;
	l: int = rq.write(out);
	cp2: RpcValue = out.read()
	print(cp1," " ,cp2," len: " ,l ," dump: " ,out);
	assert(cp1.type == cp2.type);
	rq2 = RpcRequest(cp2);
	assert(rq2.isNotify());
	assert rq2.method() == rq.method()
	assert rq2.params() == rq.params()

#if pytest doesnt work
if __name__ == "__main__":
	testRpcRequest()
	testRpcResponse()
	testRpcResponse2()
	testRpcNotify()

File: libshvchainpack/tests/test_rpcmessage.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <shv/chainpack/rpcmessage.h>
#include <shv/chainpack/chainpackwriter.h>
#include <shv/chainpack/chainpackreader.h>

#include <necrolog.h>

#include <doctest/doctest.h>

#include <optional>

using namespace shv::chainpack;
using std::string;

// Check that ChainPack has the properties we want.
#define CHECK_TRAIT(x) static_assert(std::x::value, #x)
CHECK_TRAIT(is_nothrow_constructible<RpcValue>);
CHECK_TRAIT(is_nothrow_default_constructible<RpcValue>);
CHECK_TRAIT(is_copy_constructible<RpcValue>);
CHECK_TRAIT(is_move_constructible<RpcValue>);
CHECK_TRAIT(is_copy_assignable<RpcValue>);
CHECK_TRAIT(is_move_assignable<RpcValue>);
CHECK_TRAIT(is_nothrow_destructible<RpcValue>);

namespace {

}

namespace shv::chainpack {

doctest::String toString(const RpcValue& value) {
	return value.toCpon().c_str();
}
}

DOCTEST_TEST_CASE("RpcMessage")
{

	DOCTEST_SUBCASE("RpcRequest")
	{
		RpcRequest rq;
		rq.setRequestId(123)
				.setMethod("foo")
				.setParams({{
								{"a", 45},
								{"b", "bar"},
								{"c", RpcValue::List{1,2,3}},
							}});
		rq.setMetaValue(RpcMessage::MetaType::Tag::ShvPath, "aus/mel/pres/A");
		std::stringstream out;
		RpcValue cp1 = rq.value();
		{ ChainPackWriter wr(out); rq.write(wr); }
		ChainPackReader rd(out);
		RpcValue cp2 = rd.read();
		REQUIRE(cp1.type() == cp2.type());
		RpcRequest rq2(cp2);
		REQUIRE(rq2.isRequest());
		REQUIRE(rq2.requestId() == rq.requestId());
		REQUIRE(rq2.method() == rq.method());
		REQUIRE(rq2.params() == rq.params());
	}
	DOCTEST_SUBCASE("RpcResponse result")
	{
		RpcResponse rs;
		rs.setRequestId(123).setResult(42U);
		std::stringstream out;
		RpcValue cp1 = rs.value();
		{ ChainPackWriter wr(out); rs.write(wr); }
		ChainPackReader rd(out);
		RpcValue cp2 = rd.read();
		REQUIRE(cp1.type() == cp2.type());
		RpcResponse rs2(cp2);
		REQUIRE(rs2.isResponse());
		REQUIRE(rs2.requestId() == rs.requestId());
		REQUIRE(rs2.result() == rs.result());
	}
	DOCTEST_SUBCASE("RpcResponse error")
	{
		RpcResponse rs;
		rs.setRequestId(123)
				.setError(RpcResponse::Error::create(RpcResponse::Error::InvalidParams, "Paramter length should be greater than zero!"));
		std::stringstream out;
		RpcValue cp1 = rs.value();
		{ ChainPackWriter wr(out); rs.write(wr); }
		ChainPackReader rd(out);
		RpcValue cp2 = rd.read();
		REQUIRE(cp1.type() == cp2.type());
		RpcResponse rs2(cp2);
		REQUIRE(rs2.isResponse());
		REQUIRE(rs2.requestId() == rs.requestId());
		REQUIRE(rs2.error() == rs.error());
	}
	DOCTEST_SUBCASE("RpcNotify")
	{
		RpcRequest rq;
		rq.setMethod("foo")
				.setParams({{
							   {"a", 45},
							   {"b", "bar"},
							   {"c", RpcValue::List{1,2,3}},
						   }});
		nDebug() << rq.toCpon();
		REQUIRE(rq.isSignal());
		std::stringstream out;
		RpcValue cp1 = rq.value();
		{ ChainPackWriter wr(out); rq.write(wr); }
		ChainPackReader rd(out);
		RpcValue cp2 = rd.read();
		REQUIRE(cp1.type() == cp2.type());
		RpcRequest rq2(cp2);
		REQUIRE(rq2.isSignal());
		REQUIRE(rq2.method() == rq.method());
		REQUIRE(rq2.params() == rq.params());
	}

}
